[WIP] Beautification and docs

szabo137 opened the pull request:
- [x] Fix jldoctests
- [x] finish API reference
- [x] Write tutorial for PSP
- [x] Write tutorial for stateful particles
- [x] Write tutorial for Lorentz vectors
- [x] Write tutorial for particle interface
- [x] Write tutorial for processes

AntonReinhard commented:
Looking at the documentation of the propagator, it seems outdated. It says the mass of the particle is assumed to be zero if it is not given, but `QEDcore` never accepts a mass parameter at all. I think it does (and should) get the mass from the passed particle type's `mass`. So it might be a good idea to update the docs while we're at it.

AntonReinhard commented on README.md:
```diff
@@ -1,115 +1,74 @@
-
 [![Doc Stable](https://img.shields.io/badge/docs-stable-blue.svg)](https://qedjl-project.github.io/QEDbase.jl/stable)
 [![Doc Dev](https://img.shields.io/badge/docs-dev-blue.svg)](https://qedjl-project.github.io/QEDbase.jl/dev)
 [![Build Status](https://gitlab.hzdr.de/qedjl/QEDbase.jl/badges/main/pipeline.svg)](https://gitlab.hzdr.de/qedjl/QEDbase.jl/pipelines)
 [![Coverage](https://gitlab.hzdr.de/qedjl/QEDbase.jl/badges/main/coverage.svg)](https://gitlab.hzdr.de/qedjl/QEDbase.jl/commits/main)
```
These two badges are broken... I think it's because the gitlab repositories are not public so the badges can't load.

AntonReinhard commented on README.md:
```diff
@@ -1,115 +1,74 @@
 
-This package is part of the `QuantumElectrodynamics.jl` library. For the description of the interoperability with other packages of `QuantumElectrodynamics.jl` see [docs](https://qedjl-project.github.io/QuantumElectrodynamics.jl/dev/).
+For a detailed explanation of the integration with other `QuantumElectrodynamics.jl` packages,
+please refer to the [documentation](https://qedjl-project.github.io/QuantumElectrodynamics.jl/main/).
```
This link is broken
```suggestion
please refer to the [documentation](https://qedjl-project.github.io/QuantumElectrodynamics.jl/stable/).
```

AntonReinhard commented on docs/src/_index_back.md:
Just commenting so we don't forget: This should be deleted eventually

AntonReinhard commented on docs/src/library/function_index.md:
````diff
@@ -1,5 +1,5 @@
 
 ```@index
-Pages = ["api.md", "function_index.md"]
+Pages = ["lorentz_index.md", "function_index.md"]
````
I think just the empty `@index` is enough for a list of all the functions and types.
```suggestion
```

AntonReinhard commented on README.md:
```diff
@@ -1,115 +1,74 @@
-- general Dirac bi-spinors, its adjoint counterpart as well as Dirac matrices
-- particle spinors, i.e. solutions of Dirac's equation in momentum space
-- Dirac's gamma matrices
+To lern how to implement these interfaces, please refer to the tutorials in the
```
```suggestion
To learn how to implement these interfaces, please refer to the tutorials in the
```

AntonReinhard commented on docs/make.jl:
```diff
@@ -7,40 +7,106 @@ project_path = Base.Filesystem.joinpath(Base.Filesystem.dirname(Base.source_path
+    Base.Filesystem.joinpath(project_path, "docs/src/tutorial/phase_space_point.jl"),
+]
+
+tutorial_output_dir = mktempdir(joinpath(project_path, "docs/src/"); cleanup=true)
```
While I like the idea, there is a problem with it: The generated files will live in a directory that is named differently across different builds, which means the URI will be different as well. This means that external links referencing these tutorials would have to be updated with every push to the `gh-pages` branch, or the links would be invalid. 

It doesn't seem like there is a parameter for `mktempdir` to provide a directory name though...

szabo137 replied:
You are right, of course. In 7a51195, I changed it to a fixed name `generated` and handled the garbage collections manually. Maybe we should open an issue on Julia to extend the functionality of `mktempdir` to allow fixed names.

szabo137 commented:
Since there is not much to say about models (at least not currently), I removed the respective (blank) tutorial page. Similarly, I removed the respective pages for Dirac tensors, because there is not really an interface to talk about. I suggest explaining the Dirac algebra in QEDcore, where it is fully implemented. 

From my side, I think this reached a state, where we can consider merging it.

AntonReinhard commented on docs/src/tutorial/particle_stateful.jl:
```diff
@@ -0,0 +1,94 @@
+# This method makes it easy to add more particles in the future while keeping the same structure.
+# !!! note "Reference implementation"
+#
+#     A reference implementation of a stateful particle type is given by
+#     [`QEDcore.ParticleStateful`](@extref)
```
This links to the function (the constructor), not the type. They obviously have identical names, but I think it's possible to disambiguate.

szabo137 replied:
You are right, of course. A quick internet lookup gave no insights on how to disambiguate this. Would you happen to know a way?

AntonReinhard replied:
https://juliadocs.org/DocumenterInterLinks.jl/stable/write_inventory/#The-Julia-Domain
I think something like `(@extref jl:type:QEDcore.ParticleStateful)` should work but I haven't tried it myself yet.

szabo137 replied:
Thanks for the suggestion. Done in 97e4caa

AntonReinhard reviewed (commented):
Just one small broken link
